File: Task5.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <algorithm>
#include <ctime>
#include <cstdlib>
using namespace std;

vector<vector<int>> generateRandomBoard(int rows, int cols) {
    vector<vector<int>> board(rows, vector<int>(cols));
    srand(time(0));
    
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            board[i][j] = rand() % 2;
        }
    }
    return board;
}

void printBoard(const vector<vector<int>>& board) {
    for (const auto& row : board) {
        for (int coin : row) {
            cout << coin << " ";
        }
        cout << endl;
    }
}

int countMismatches(const vector<vector<int>>& board, bool startWith) {
    int mismatches = 0;
    bool expected = startWith;
    for (const auto& row : board) {
        bool currentExpected = expected;
        for (int coin : row) {
            if (coin != currentExpected) {
                mismatches++;
            }
            currentExpected = !currentExpected;
        }
        expected = !expected;
    }
    return mismatches;
}

int solveCoinBoard(const vector<vector<int>>& board) {
    // Подсчет несоответствий для двух возможных шахматных паттернов
    int mismatches1 = countMismatches(board, true);
    int mismatches2 = countMismatches(board, false);
    
    if (mismatches1 % 2 != 0 && mismatches2 % 2 != 0) {
        return -1;
    }
    
    // Минимальное количество ходов - половина от несоответствий (так как переворачиваем по 2 монеты)
    int minMoves1 = (mismatches1 % 2 == 0) ? mismatches1 / 2 : INT_MAX;
    int minMoves2 = (mismatches2 % 2 == 0) ? mismatches2 / 2 : INT_MAX;
    
    if (minMoves1 == INT_MAX && minMoves2 == INT_MAX) {
        return -1;
    }
    return min(minMoves1, minMoves2);
}

int main() {
    int rows, cols;
    cout << "Введите количество строк и столбцов доски: ";
    cin >> rows >> cols;
    
    if (rows <= 0 || cols <= 0) {
        cout << "Некорректные размеры доски!" << endl;
        return 1;
    }
    
    vector<vector<int>> board = generateRandomBoard(rows, cols);
    
    cout << "\nСгенерированная доска:\n";
    printBoard(board);
    
    int minMoves = solveCoinBoard(board);
    
    if (minMoves == -1) {
        cout << "\nНевозможно преобразовать доску в шахматный порядок";
    } else {
        cout << "\nМинимальное количество ходов: " << minMoves << endl;
    }
    return 0;
}

File: Task5.rs
use rand::Rng;
use std::cmp::min;
use std::i32::MAX;

fn generate_random_board(rows: usize, cols: usize) -> Vec<Vec<i32>> {
    let mut rng = rand::thread_rng();
    (0..rows)
        .map(|_| (0..cols).map(|_| rng.gen_range(0..2)).collect())
        .collect()
}

fn print_board(board: &[Vec<i32>]) {
    for row in board {
        for coin in row {
            print!("{} ", coin);
        }
        println!();
    }
}

fn count_mismatches(board: &[Vec<i32>], start_with: bool) -> i32 {
    let mut mismatches = 0;
    let mut expected = start_with;
    
    for row in board {
        let mut current_expected = expected;
        for &coin in row {
            if (coin == 1) != current_expected {
                mismatches += 1;
            }
            current_expected = !current_expected;
        }
        expected = !expected;
    }
    mismatches
}

fn solve_coin_board(board: &[Vec<i32>]) -> i32 {
    let mismatches1 = count_mismatches(board, true);
    let mismatches2 = count_mismatches(board, false);
    
    if mismatches1 % 2 != 0 && mismatches2 % 2 != 0 {
        return -1;
    }
    
    let min_moves1 = if mismatches1 % 2 == 0 {
        mismatches1 / 2
    } else {
        MAX
    };
    
    let min_moves2 = if mismatches2 % 2 == 0 {
        mismatches2 / 2
    } else {
        MAX
    };
    
    if min_moves1 == MAX && min_moves2 == MAX {
        -1
    } else {
        min(min_moves1, min_moves2)
    }
}

fn main() {
    println!("Введите количество строк и столбцов доски: ");
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).expect("Ошибка чтения ввода");
    
    let dimensions: Vec<usize> = input
        .trim()
        .split_whitespace()
        .map(|s| s.parse().unwrap_or(0))
        .collect();
    
    if dimensions.len() != 2 || dimensions[0] == 0 || dimensions[1] == 0 {
        println!("Некорректные размеры доски!");
        return;
    }
    
    let (rows, cols) = (dimensions[0], dimensions[1]);
    let board = generate_random_board(rows, cols);
    
    println!("\nСгенерированная доска:");
    print_board(&board);
    
    let min_moves = solve_coin_board(&board);
    
    if min_moves == -1 {
        println!("\nНевозможно преобразовать доску в шахматный порядок");
    } else {
        println!("\nМинимальное количество ходов: {}", min_moves);
    }
}
